vochain/app.go

```diff
@@ -285,27 +298,22 @@ func (app *BaseApplication) InitChain(req abcitypes.RequestInitChain) abcitypes.
 // punishments for the validators.
 func (app *BaseApplication) BeginBlock(
 	req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
-	header := &models.TendermintHeader{
-		Height:         req.Header.GetHeight(),
-		ConsensusHash:  req.Header.GetConsensusHash(),
-		AppHash:        req.Header.GetAppHash(),
-		BlockID:        req.Header.LastBlockId.GetHash(),
-		DataHash:       req.Header.GetDataHash(),
-		LastCommitHash: req.Header.GetLastCommitHash(),
-		Timestamp:      req.Header.GetTime().Unix(),
-	}
-	headerBytes, err := proto.Marshal(header)
-	if err != nil {
-		log.Warnf("cannot marshal header in BeginBlock")
-	}
-	// reset app state to latest persistent data
 	app.State.Rollback()
-	app.State.Lock()
-	if err = app.State.Tx.Set(headerKey, headerBytes); err != nil {
-		log.Fatal(err)
-	}
-	app.State.Unlock()
-	go app.State.CachePurge(uint32(header.Height))
+	atomic.StoreInt64(&app.startBlockTimestamp, req.Header.GetTime().Unix())
+	height := uint32(req.Header.GetHeight())
+	app.State.SetHeight(height)
+	// NOTE(Edu): I think there's a race here.  CachePurge reads elements
+	// of the State.voteCache while there are other goroutines that write
+	// to elements in State.voteCache (for example
+	// BaseApplication.VoteEnvelopeCheck calls State.CacheGet and modifies
+	// the returned element).  I think a possible solution would be for
+	// State.CacheGet to return a copy of the element, but the returned
+	// type is protobuf generated and is not copy friendly.  Can we create
+	// a new element manually copying the fields?  (it's cumbersome and can
+	// be error prone if more fields are added in the future).  Can we find
+	// another solution?  @mvdan got any idea?  At the end I decided to use
+	// proto.Clone which seems appropiate.  Please comment!
```

Comment:
I think your spotted race is right. We probably haven't found it because none of our tests end up purging the cache while VoteEnvelopeCheck runs.

Your solution seems right, it's just slightly inefficient. We're going to do a full generic copy at every Get, even for the cases that don't need it, like CacheDel.

An alternative is to make both ends use sync/atomic, since the field is just a uint32, and we only seem to access it in two places. That would definitely be cheaper, with the downside that we always need to remember to use atomics when accessing it. That actually seems like a problem, because then any calls to `proto.Marshal` or `proto.Unmarshal` would be unsafe since they would not use atomics.

There's also a middle ground: use `proto.Clone`, but only when we're about to modify a vote struct: in VoteEnvelopeCheck. There you can explain why we need it.

Reply:
There's also a subtle behavior change here: in master right now, VoteEnvelopeCheck updates the height of the vote in the cache, so other goroutines could see that update in-place. If we clone, then that would never happen.

VoteEnvelopeCheck seems to do a CacheDel right after modifying the vote, so I think we are not meant to share this update. So, in that sense, your Clone is more correct than my atomics, since my atomics could still lead to other goroutines seeing the update when they don't want to.